### PR TITLE
Spaces Swift SDK example app

### PR DESCRIPTION
**Pull request by andrewjl-mux**
- add Spaces Swift SDK example app
- update .gitignore exclusions

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ Get to know Mux by looking through examples! Contributions are always welcome, e
 | [swift-video-app](swift-video-app)                      | Swift app that uses Mux Video, Mux Data and plays audio when the app goes into the background | VOD, Mobile | Swift |
 | [aws-recommendation-engine](aws-recommendation-engine)  | An example of using Mux Data streaming exports to recommend content with Amazon Personalize | VOD, Serverless | TypeScript |
 | [nextjs-mdx-player](nextjs-mdx-player)                  | A [Next.js](https://nextjs.org/) app configured to support [MDX](https://mdxjs.com) to illustrate supporting [MuxPlayer](https://www.mux.com/player) in blog posts. | VOD, CMS | JavaScript |
+| [swift-spaces-real-time-app](swift-spaces-real-time-app)| An example iOS and iPadOS app demonstrating usage of the Spaces Swift SDk for real-time video
```
Comment (refactornator) [inline README.md:20]: Probably good to add tags and language to this.

File: swift-spaces-real-time-app/README.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+## Installation
+
+### Adding the SDK using Swift Package Manager
+
+You can add the SDK using Swift Package Manager via Xcode by following the steps described (here)[https://developer.apple.com/documentation/xcode/adding-package-dependencies-to-your-app]. Use this [Swift Package URL](https://github.com/muxinc/mux-spaces-sdk-swift-distribution) when adding the SDK in Xcode.
+
+### Adding the SDK manually
+
+Follow the steps here to add the SDK manually.
```
Comment (refactornator) [inline swift-spaces-real-time-app/README.md:15]: It's unclear which steps to follow here.

Comment (andrewjl-mux) [reply swift-spaces-real-time-app/README.md]: I've removed this section.

File: swift-spaces-real-time-app/MuxSpacesUIKit/Space+Token.swift [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+/// Set the Mux Spaces token as an environment variable to try
+/// out the app. See [here](https://developer.apple.com/documentation/xcode/customizing-the-build-schemes-for-a-project?changes=_8) for step-by-step directions.
+let token = ProcessInfo.processInfo.environment[
+    "MUX_SPACES_TOKEN"
+] ?? ""
+
+extension Space {
+    static func make() throws -> Space {
+        if token.isEmpty {
+            print("Please add a include Spaces JWT to try out the sample code")
```
Comment (refactornator) [inline swift-spaces-real-time-app/MuxSpacesUIKit/Space+Token.swift:17]:
```suggestion
            print("Please include a Spaces JWT to try out the sample code")
```

File: swift-spaces-real-time-app/MuxSpacesUIKit/View/JoinSpaceViewController.swift [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+    var videoToggle: UISwitch = UISwitch()
+    var videoStackView: UIStackView = UIStackView()
+
+    var captureDeviceSelectionLabel: UILabel = UILabel()
+    var captureDeviceSelectionControl: UISegmentedControl = UISegmentedControl()
+
+    var cancellables: [AnyCancellable] = []
+
+    @Published var isAudioPublishingEnabled: Bool = false
+    @Published var isVideoPublishingEnabled: Bool = false
```
Comment (refactornator) [inline swift-spaces-real-time-app/MuxSpacesUIKit/View/JoinSpaceViewController.swift:37]:
```suggestion
    @Published var isAudioPublishingEnabled: Bool = true
    @Published var isVideoPublishingEnabled: Bool = true
```

Comment (refactornator) [reply swift-spaces-real-time-app/MuxSpacesUIKit/View/JoinSpaceViewController.swift:37]: I think it makes more sense to default this to enabled so audio and video are published on joining.

File: swift-spaces-real-time-app/MuxSpacesUIKit/View/JoinSpaceViewController.swift [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+        audioLabel.text = NSLocalizedString(
+            "Publish Audio?",
+            comment: "Publish Audio Toggle Label"
+        )
+        audioLabel.font = UIFont.preferredFont(
+            forTextStyle: .callout
+        ).withSize(16.0)
+        audioLabel.textColor = .label
+        audioToggle.translatesAutoresizingMaskIntoConstraints = false
+        audioToggle.isEnabled = true
```
Comment (refactornator) [inline swift-spaces-real-time-app/MuxSpacesUIKit/View/JoinSpaceViewController.swift:106]:
```suggestion
        audioToggle.isEnabled = true
        audioToggle.isOn = true
```

File: swift-spaces-real-time-app/MuxSpacesUIKit/View/JoinSpaceViewController.swift [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+        videoLabel.text = NSLocalizedString(
+            "Publish Video?",
+            comment: "Publish Video Toggle Label"
+        )
+        videoLabel.font = UIFont.preferredFont(
+            forTextStyle: .callout
+        ).withSize(16.0)
+        videoLabel.textColor = .label
+        videoToggle.translatesAutoresizingMaskIntoConstraints = false
+        videoToggle.isEnabled = true
```
Comment (refactornator) [inline swift-spaces-real-time-app/MuxSpacesUIKit/View/JoinSpaceViewController.swift:150]:
```suggestion
        videoToggle.isEnabled = true
        videoToggle.isOn = true
```

**Review by refactornator — APPROVED** [review]
I got the sample app to run on my phone. I made a few suggestions like defaulting to publishing audio and video. As well as a wording fix. The sample app works, so I think we're good to merge it and we can continue to improve it.

One surprising thing was that I had to "reset package cache" for it to find the MuxSpaces dependency and not fail. I'm not sure if there's any way to prevent that from needing to be reset.

Screen Shot 2022-10-28 at 7 09 22 PM

**Comment by andrewjl-mux** [comment]
> I got the sample app to run on my phone. I made a few suggestions like defaulting to publishing audio and video. As well as a wording fix. The sample app works, so I think we're good to merge it and we can continue to improve it.

Thanks for flagging these! The most recent changes improve the state handling further and these publishing audio and video flags ended up being duplicate state, so they were removed. The spaces token is now also injected as an environment variable.

> One surprising thing was that I had to "reset package cache" for it to find the MuxSpaces dependency and not fail. I'm not sure if there's any way to prevent that from needing to be reset.

Good catch, fixed this by removing the Swift package from the Xcode project. This does mean that the SDK needs to be installed first for the app to be usable. I updated the app README to highlight this.

**Comment by refactornator** [comment]
I pulled the latest changes and added the MuxSpaces package from SPM and successfully ran the app 🎉

One odd inconsistency is that when I open the app, the toggles for publishing audio and video look off, but the app shows the audio and video options I get when they are toggled on. I think the internal state is that they are on. Maybe we should just remove the toggles and always publish audio and video? For a sample app, I would be ok with this, if you agree.
IMG_6360

I was able to successfully join a space and everything seemed to be working there. But I noticed that if I left the space, and immediately tried to rejoin, I got an error that my JWT was invalid. This might be annoying for anyone testing with the sample app. The JWT should still be valid and it should be possible to rejoin the space.

**Comment by andrewjl-mux** [comment]
Both the toggles and re-joining the space have been fixed, please keep the feedback coming and I will address in a follow-on PR.

**Pull request merged**